lib/SyTest/Homeserver/ProcessManager.pm

```diff
@@ -237,5 +238,129 @@ sub _kill_process
    );
 }
 
-1;
+=head2 _start_process_and_await_notify
+
+   $fut = $hs->_start_process_and_await_notify( %params )
+
+This method starts a new process, setting the `NOTIFY_SOCKET` env, and waits to
+receive a `READY=1` notification from the process on the socket.
+
+Parameters are passed on to C<IO::Async::Process::new>.
+
+=cut
+
+sub _start_process_and_await_notify
+{
+   my $self = shift;
+
+   my %params = @_;
+
+   # We now need to do some faffing to pull out any passed env hash so that we
+   # can then pass it to `_await_ready_notification`.
+   #
+   # The env is passed in as part of the `setup` param, which is an ordered map
+   # represented as an array.
+   $params{setup} //= [];
+
+   my $setup = $params{setup};
+
+   my %setup_map = @$setup;  # Copy to a hash so we can pull out the env entry.
+   my $env = $setup_map{env} // {};
+   if ( not defined $setup_map{env} ) {
+      # There was no env entry, so we me need to add it to the setup array.
+      push @$setup, env => $env;
+   }
+
+   # We need to set this up *before* we start the process as we need to bind the
+   # socket before startin the process.
+   my $await_fut = $self->_await_ready_notification( $env );
+
+   my $proc = $self -> _start_process( %params );
+   my $finished_future = $self->{proc_info}{$proc}->finished_future;
+
+   my $fut = Future->wait_any(
+      $await_fut,
+      $finished_future->without_cancel()->then_fail(
+         "Process died without becoming connectable",
+      ),
+   )->else_with_f( sub {
+      my ( $f ) = @_;
+
+      # We need to manually kill child procs here as we don't seem to have
+      # registered the on finish handler yet.
+      $self->kill_and_await_finish()->then( sub {
+         $f
+      })
+   } );
```

Comment:
this feels wrong, for a couple of reasons:
* failure to start one subprocess should not necessarily mean that all other subprocesses get killed (at least at the abstract level of a ProcessManager class)
* it's asymmetric with `_start_process_and_await_connectable`.

what is the "finish handler" of which the comment speaks?

Reply:
Ah, turns out that there is a timeout several layers up. Its set to 60s so I probably got bored waiting before it fired :/